lib/SILOptimizer/Utils/Existential.cpp

```diff
@@ -311,3 +312,83 @@ ConcreteExistentialInfo::ConcreteExistentialInfo(Operand &openedUse) {
   }
   assert(isValid());
 }
+
+/// Initialize a ConcreteExistentialInfo based on the already computed concrete
+/// type and protocol declaration. It first determines the OpenedArchetypeDef
+/// from the Arg argument and then appends new unchecked_cast instructions in
+/// the function body using the SILBuilder.
+ConcreteExistentialInfo::ConcreteExistentialInfo(Operand &ArgOperand,
+                                                 const CanType &ConcreteTy,
+                                                 ProtocolDecl *Protocol,
+                                                 SILBuilder &Builder)
+    : ConcreteType(ConcreteTy) {
+  SILValue Arg = ArgOperand.get();
+  SILInstruction *User = ArgOperand.getUser();
+
+  // We need the OpenedArchetypeDef to be able to cast it to the concrete type.
+  // findInitExistential helps us determine this OpenedArchetypeDef. For cases
+  // where OpenedArchetypeDef can not be found from findInitExistential (because
+  // there was no InitExistential), then the code after findInitExistential does
+  // extra effort in trying to find an OpenedArchetypeDef for AllocStackInst
+  // using getAddressOfStackInit.
+  isCopied = false;
+  OpenedArchetypeDef = Arg;
+  InitExistential = findInitExistential(ArgOperand, OpenedArchetype,
+                                        OpenedArchetypeDef, isCopied);
+
+  // If findInitExistential did not update the OpenedArchetypeDef (because it
+  // did not find an InitExistential) and that the original Arg is an
+  // alloc_stack instruction, then we determine the OpenedArchetypeDef using
+  // getAddressOfStackInit. Please keep in mind that an alloc_stack can be
+  // an argument to apply (wich has no InitExistential), thus we need to
+  // determine the OpenedArchetypeDef for it. This is the extra effort.
+  if (auto *Instance = dyn_cast<AllocStackInst>(OpenedArchetypeDef)) {
+    bool copied = false;
+    if (SILValue Src = getAddressOfStackInit(Instance, User, copied)) {
+      OpenedArchetypeDef = Src;
+    }
+    isCopied = copied;
+  }
+
+  // Bail, if we did not find an opened existential.
+  if (!(isa<OpenExistentialRefInst>(OpenedArchetypeDef) ||
+        isa<OpenExistentialAddrInst>(OpenedArchetypeDef)))
+    return;
+
+  // Create SIL type for the concrete type.
+  auto &M = Builder.getModule();
+  SILType ConcreteSILType = M.Types.getLoweredType(ConcreteType);
+
+  // We have the open_existential; we still need the conformance.
+  auto ConformanceRef =
+      M.getSwiftModule()->lookupConformance(ConcreteType, Protocol);
```

Comment:
We're generally not supposed to call `SwiftModule::lookupConformance` from the optimizer because the conformance hasn't necessarilly been type checked. In this case, I don't really see an alternative though. @slavapestov , do you have any suggestions? Are there certain conditions where we can say that it's safe to do this?